### ISSUE TITLE
How to set default ksmps via libcsound

### ISSUE THREAD

**Issue by giuliomoro** [csound/csound#1444]
When running Csound on a platform with relatively low computing power and small blocksizes and "single buffering" of the audio (i.e.: no input/output FIFO as provided, for instance, by ALSA), using the default value of ksmps=10 can be undesirable. One would want to ensure that an integer and constant number of `ksmps` is contained in the audio block size, in order to ensure that the computational cost of the audio callback is as constant as possible (with usual variations due to scheduling, caching and non-constant algorithm complexity).

I would like to programmatically set a default for the user of the program so that ksmps defaults to the same as the audio blocksize, but also allow them to override that in any of the usual ways. Note that the blocksize is a runtime parameter, so simply redefining `DFLT_KSMPS` wouldn' work.

Is there a way to set a default value for `ksmps` when initialising libcsound before loading the program?
Basically I would like something like this:

```
    csound = new Csound();
    csound->SetHostData((void*)context);
    csound->SetHostImplementedAudioIO(1, 0);
    csound->SetHostImplementedMIDIIO(1);
   // DO MAGIC HERE: set default ksmps to context->audioSamples
    const char* args[] = { "csound",  csdfile.c_str(), "-iadc", "-odac", "-+rtaudio=null", "--realtime",
                           "--daemon" };
    int numArgs = (int)(sizeof(args) / sizeof(char*));
    csound->Compile(numArgs, args); // load the `csd` file, which may override ksmps
```

How would you implement `DO MAGIC` where indicated (or elsewhere)?

Thanks

**Comment by jpffitch** [comment]
Not at a real computer at present but there is an option to override or or ksmps.  Check the manual......

⁣Sent from TypeApp ​

**Comment by vlazzarini** [comment]
Hi Giulio,

you can use setOption("--ksmps=N"), where N id your desired value.

or since you are already setting other flags, you can add the option to that string.

**Comment by giuliomoro** [comment]
Thanks Victor, however neither of those seem to allow the user to override ksmps by setting for instance

```
<CsInstruments>
ksmps = 100
```

So for instance

```
    Csound* csound = new Csound();
    csound->SetHostData((void*)context);
    csound->SetHostImplementedAudioIO(1, 0);
    csound->SetHostImplementedMIDIIO(1);
    const char* args[] = { "csound",  csdfile.c_str(), "-iadc", "-odac", "-+rtaudio=null", "--realtime",
                           "--daemon" };
    int numArgs = (int)(sizeof(args) / sizeof(char*));
    csound->SetOption("--ksmps=16");
    csound->Compile(numArgs, args);
    printf(">>> %d\n", csound->GetKsmps());
    return false;
```
always prints  (among other things)
```
>>> 16
```

regardless of the value of 
```
<CsInstruments>
ksmps = 100
```
in the csd file loaded with `Compile` file. Is there a way to have the latter override the default set by `SetOption()` ?

**Comment by rorywalsh** [comment]
Hi Giuliormo. You can use the CSOUND_PARAMS struct. It has ksmps_override
member. Then pass the struct to Csound::SetParam(). That should do it I
think.

**Comment by vlazzarini** [comment]
Hi Giulio,

if you set ksmps = 16 it will override whatever the user sets.

```
  csound->SetOption("--ksmps=16");
```

will fix it to 16. That takes precedence over whatever the user sets The precedence is

- option overrides
- ksmps set by the user

It is a bit of a chicken and egg situation, as the CSD comes after your default setting.

**Comment by giuliomoro** [comment]
> You can use the CSOUND_PARAMS struct. It has ksmps_override member. Then pass the struct to Csound::SetParam().

that seems to have the same result as above.

> The precedence is
> 
> * option overrides
> * ksmps set by the user

and this confirms that the observed behaviour is the expected behaviour.

Let's try to look at the problem for a different angle: is it possible to do a quick "compile" sort of thing where only the options (or at least ksmps) set in the csd file are parsed and made available? That way I could do that first, check whether  a `ksmps` is actually set by the user, in which case I'd avoid setting an "option override" before calling `Compile()`.

Failing that, I could probably call `Compile()` twice, where the purpose of the first call is uniquely to call `GetKsmps()` after it, but this seems like it would be more wasteful. Is `Compile()` going to be slow in some cases (on embedded platform!)? Is there a way to "undo" a `Compile()` without destroying the `Csound` object?

**Comment by vlazzarini** [comment]
The issue is that once you call compile, the ksmps is set until there is an object reset. Besides thi there is no way to know whether the ksmps was set by the user, unless you call csoundParseOrc() and examine the tree. The snag is that that operates on an orchestra string rather than the CSD, since that's the only function currently exposed in the API.

Is there a configuration window you could provide where ksmps is always set and which always overrides the CSD?
Another idea is to parse the CSD and look for "ksmps = ".

Possibly we could think of a configuration variable to set the default ksmps before compilation. What do you reckon @jpff?

**Comment by vlazzarini** [comment]
I'm looking at setting defaults as configuration parameters, which we could then offer as options.

**Comment by vlazzarini** [comment]
Ok: a new option is there "--default-ksmps=N", which only overrides the default ksmps (10). That should do the job for you.

**Comment by giuliomoro** [comment]
thanks a lot, that works great!
